File: packages/platform-go/money/money.go
// Package money defines the platform Money value type shared by Go
// services. Mirrors the Kotlin `Money` (@JvmInline value class) in
// `platform-spring-boot-money` and the Python `Money` model in
// `platform-python`. Wire JSON: `{"amount": "19.99", "currency": "USD"}`.
// All arithmetic is performed on minor units (Long / int64) to avoid
// floating-point drift. Mixed-currency arithmetic is rejected at
// runtime.
package money

import (
	"errors"
	"fmt"
	"math/big"
)

// Money is a value type carrying `minor` (Long minor units) and
// `currency` (ISO 4217 code).
type Money struct {
	Minor    int64  `json:"-"`
	Currency string `json:"-"`
}

// Amount is the display-form representation, used for wire JSON.
type Amount struct {
	Amount   string `json:"amount"`
	Currency string `json:"currency"`
}

// Zero returns the zero amount for a given currency.
func Zero(currency string) Money { return Money{Minor: 0, Currency: currency} }

// OfMinor constructs a Money from minor units (e.g. cents).
func OfMinor(minor int64, currency string) Money { return Money{Minor: minor, Currency: currency} }

// Of constructs a Money from a display-form amount string (e.g. "19.99").
// The number of fractional digits is determined by the ISO 4217 currency.
func Of(amount string, currency string) (Money, error) {
	rat, ok := new(big.Rat).SetString(amount)
	if !ok {
		return Money{}, fmt.Errorf("money: invalid amount %q", amount)
	}
	frac := defaultFractionDigits(currency)
	scale := new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(frac)), nil)
	rat.Mul(rat, new(big.Rat).SetInt(scale))
	minor := new(big.Int).Div(rat.Num(), rat.Denom())
	if !rat.IsInt() {
		// rounded to nearest
		num2 := new(big.Int).Mul(rat.Num(), big.NewInt(2))
		denom2 := rat.Denom()
		half := new(big.Int).Div(denom2, big.NewInt(2))
		if num2.Cmp(new(big.Int).Mul(denom2, minor)) >= 0 {
			// round up
			mod := new(big.Int).Mod(num2, denom2)
			if mod.Cmp(zero) != 0 {
				minor.Add(minor, big.NewInt(1))
			}
		}
		_ = half
	}
	if !minor.IsInt64() {
		return Money{}, errors.New("money: minor overflow")
	}
	return Money{Minor: minor.Int64(), Currency: currency}, nil
}

// Major returns the display-form big.Rat (e.g. 19.99).
func (m Money) Major() string {
	frac := defaultFractionDigits(m.Currency)
	scale := new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(frac)), nil)
	num := new(big.Int).Mul(big.NewInt(m.Minor), big.NewInt(1))
	rat := new(big.Rat).SetFrac(num, scale)
	return rat.FloatString(frac)
}

// UnmarshalJSON parses the wire format {"amount": "...", "currency": "..."}.
func (m *Money) UnmarshalJSON(data []byte) error {
	var a Amount
	if err := jsonUnmarshal(data, &a); err != nil {
		return err
	}
	parsed, err := Of(a.Amount, a.Currency)
	if err != nil {
		return err
	}
	*m = parsed
	return nil
}

// MarshalJSON renders the wire format.
func (m Money) MarshalJSON() ([]byte, error) {
	return jsonMarshal(Amount{Amount: m.Major(), Currency: m.Currency})
}

func (m Money) Plus(other Money) (Money, error) {
	if m.Currency != other.Currency {
		return Money{}, fmt.Errorf("money: mixed currencies %s + %s", m.Currency, other.Currency)
	}
	return Money{Minor: m.Minor + other.Minor, Currency: m.Currency}, nil
}

func (m Money) Minus(other Money) (Money, error) {
	if m.Currency != other.Currency {
		return Money{}, fmt.Errorf("money: mixed currencies %s - %s", m.Currency, other.Currency)
	}
	return Money{Minor: m.Minor - other.Minor, Currency: m.Currency}, nil
}

func (m Money) Times(multiplier int64) Money {
	return Money{Minor: m.Minor * multiplier, Currency: m.Currency}
}

func (m Money) Div(divisor int64) Money {
	return Money{Minor: m.Minor / divisor, Currency: m.Currency}
}

func defaultFractionDigits(currency string) int {
	switch currency {
	case "JPY", "KRW", "VND", "CLP", "PYG", "UGX", "XAF", "XOF":
		return 0
	case "BHD", "JOD", "KWD", "OMR", "TND":
		return 3
	default:
		return 2
	}
}

var zero = big.NewInt(0)

// jsonUnmarshal and jsonMarshal are split out so the dependency on
// encoding/json doesn't bleed into a file that is mostly arithmetic.
func jsonUnmarshal(data []byte, v any) error { return jsonImpl.Unmarshal(data, v) }
func jsonMarshal(v any) ([]byte, error)       { return jsonImpl.Marshal(v) }

File: packages/platform-python/src/platform_python/money.py
"""Money value type shared by Python services.

Wire JSON: ``{"amount": "19.99", "currency": "USD"}``. Arithmetic is
performed on minor units (Long / int64) to avoid floating-point drift.
Mixed-currency arithmetic is rejected at runtime.
"""
from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal, getcontext
from typing import Annotated

from pydantic import BaseModel, ConfigDict, Field, model_validator

# Default precision: 28 digits is more than enough for any ISO 4217 amount.
getcontext().prec = 28

# ISO 4217 currencies with 0 or 3 fractional digits.
_ZERO_FRACTION_CURRENCIES = frozenset({"JPY", "KRW", "VND", "CLP", "PYG", "UGX", "XAF", "XOF"})
_THREE_FRACTION_CURRENCIES = frozenset({"BHD", "JOD", "KWD", "OMR", "TND"})


def _fraction_digits(currency: str) -> int:
    if currency in _ZERO_FRACTION_CURRENCIES:
        return 0
    if currency in _THREE_FRACTION_CURRENCIES:
        return 3
    return 2


class Money(BaseModel):
    """Value type holding minor units + ISO 4217 currency."""

    model_config = ConfigDict(frozen=True)

    minor: int
    currency: Annotated[str, Field(min_length=3, max_length=3)]

    @classmethod
    def zero(cls, currency: str) -> "Money":
        return cls(minor=0, currency=currency)

    @classmethod
    def of_minor(cls, minor: int, currency: str) -> "Money":
        return cls(minor=minor, currency=currency)

    @classmethod
    def of(cls, amount: str | Decimal, currency: str) -> "Money":
        if isinstance(amount, str):
            amount = Decimal(amount)
        frac = _fraction_digits(currency)
        scale = Decimal(10) ** frac
        quant = (amount * scale).quantize(Decimal("1"), rounding=ROUND_HALF_UP)
        return cls(minor=int(quant), currency=currency)

    @property
    def major(self) -> Decimal:
        frac = _fraction_digits(self.currency)
        scale = Decimal(10) ** frac
        return Decimal(self.minor) / scale

    def plus(self, other: "Money") -> "Money":
        self._assert_same_currency(other)
        return Money(minor=self.minor + other.minor, currency=self.currency)

    def minus(self, other: "Money") -> "Money":
        self._assert_same_currency(other)
        return Money(minor=self.minor - other.minor, currency=self.currency)

    def times(self, multiplier: int) -> "Money":
        return Money(minor=self.minor * multiplier, currency=self.currency)

    def div(self, divisor: int) -> "Money":
        return Money(minor=self.minor // divisor, currency=self.currency)

    def _assert_same_currency(self, other: "Money") -> None:
        if self.currency != other.currency:
            raise ValueError(f"money: mixed currencies {self.currency} and {other.currency}")

    @model_validator(mode="before")
    @classmethod
    def _from_wire(cls, data: Any) -> Any:
        """Accept wire format ``{"amount": "...", "currency": "..."}`` or the
        canonical ``{"minor": ..., "currency": ...}`` shape."""
        if isinstance(data, dict) and "amount" in data and "minor" not in data:
            amount = data["amount"]
            currency = data["currency"]
            return cls.of(amount, currency).model_dump()
        return data


class MoneyHTTPException(Exception):
    """Convenience exception carrying a fully built Money for HTTP responses."""

    def __init__(self, money: Money) -> None:
        self.money = money
        super().__init__(f"{money.major} {money.currency}")
